File: coolercontrol/services/liquidctl_device_extractors/kraken_2.py
#  CoolerControl - monitor and control your cooling and other devices
#  Copyright (c) 2022  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging
from typing import List, Dict, Any

from liquidctl.driver import kraken2
from liquidctl.driver.kraken2 import Kraken2

from coolercontrol.models.channel_info import ChannelInfo
from coolercontrol.models.device_info import DeviceInfo
from coolercontrol.models.lighting_mode import LightingMode
from coolercontrol.models.speed_options import SpeedOptions
from coolercontrol.models.status import TempStatus, ChannelStatus
from coolercontrol.services.liquidctl_device_extractors import LiquidctlDeviceInfoExtractor

_LOG = logging.getLogger(__name__)


# pylint: disable=protected-access
class Kraken2Extractor(LiquidctlDeviceInfoExtractor):
    supported_driver = Kraken2
    _channels: Dict[str, ChannelInfo] = {}
    _lighting_speeds: List[str] = []
    _min_liquid_temp = 20
    _max_liquid_temp = 60
    _backwards_enabled_modes: List[str] = [
        'spectrum-wave',
        'super-wave',
        'marquee-3',
        'marquee-4',
        'marquee-5',
        'marquee-6',
        'covering-marquee',
        'moving-alternating'
    ]
    _speed_disabled_modes: List[str] = [
        'off',
        'fixed',
        'super-fixed'
    ]

    @classmethod
    def extract_info(cls, device_instance: Kraken2) -> DeviceInfo:
        if device_instance.supports_cooling:
            for channel_name, (_, duty_min, duty_max) in kraken2._SPEED_CHANNELS.items():
                cls._channels[channel_name] = ChannelInfo(
                    speed_options=SpeedOptions(
                        min_duty=duty_min,
                        max_duty=duty_max,
                        profiles_enabled=device_instance.supports_cooling_profiles,
                        fixed_enabled=device_instance.supports_cooling
                    )
                )

        for channel_name in kraken2._COLOR_CHANNELS:
            cls._channels[channel_name] = ChannelInfo(
                lighting_modes=cls._get_filtered_color_channel_modes(channel_name)
            )

        cls._lighting_speeds = list(kraken2._ANIMATION_SPEEDS.keys())

        return DeviceInfo(
            channels=cls._channels,
            lighting_speeds=cls._lighting_speeds,
            temp_min=cls._min_liquid_temp,
            temp_max=cls._max_liquid_temp,
            profile_max_length=9,
            temp_ext_available=True
        )

    @classmethod
    def _get_filtered_color_channel_modes(cls, channel_name: str) -> List[LightingMode]:
        channel_modes: List[LightingMode] = []
        for mode_name, (_, _, _, min_colors, max_colors, only_ring) in kraken2._COLOR_MODES.items():
            if not only_ring or only_ring and channel_name == 'ring':
                if 'backwards' not in mode_name:  # remove any deprecated modes
                    channel_modes.append(LightingMode(
                        mode_name, cls._channel_to_frontend_name(mode_name),
                        min_colors, max_colors,
                        mode_name not in cls._speed_disabled_modes,
                        mode_name in cls._backwards_enabled_modes
                    ))
        return channel_modes

    @classmethod
    def _get_temperatures(cls, status_dict: Dict[str, Any], device_id: int) -> List[TempStatus]:
        temps = []
        liquid = cls._get_liquid_temp(status_dict)
        if liquid is not None:
            temps.append(TempStatus('liquid', liquid, 'Liquid', f'LC#{device_id} Liquid'))
        return temps

    @classmethod
    def _get_channel_statuses(cls, status_dict: Dict[str, Any]) -> List[ChannelStatus]:
        channel_statuses: List[ChannelStatus] = []
        fan_rpm = cls._get_fan_rpm(status_dict)
        if fan_rpm is not None:
            channel_statuses.append(ChannelStatus('fan', rpm=fan_rpm))
        pump_rpm = cls._get_pump_rpm(status_dict)
        if pump_rpm is not None:
            channel_statuses.append(ChannelStatus('pump', rpm=pump_rpm))
        return channel_statuses

File: coolercontrold/src/repositories/liquidctl/supported_devices/kraken2.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2021-2024  Guy Boldon, Eren Simsek and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::collections::HashMap;

use crate::device::{ChannelInfo, DeviceInfo, LightingMode, SpeedOptions};
use crate::repositories::liquidctl::base_driver::BaseDriver;
use crate::repositories::liquidctl::liqctld_client::DeviceProperties;
use crate::repositories::liquidctl::supported_devices::device_support::{ColorMode, DeviceSupport};

#[derive(Debug)]
pub struct Kraken2Support;

impl Kraken2Support {
    pub fn new() -> Self {
        Self {}
    }
}

impl DeviceSupport for Kraken2Support {
    fn supported_driver(&self) -> BaseDriver {
        BaseDriver::Kraken2
    }

    fn extract_info(&self, _device_index: &u8, device_props: &DeviceProperties) -> DeviceInfo {
        let mut channels = HashMap::new();
        if let Some(true) = device_props.supports_cooling {
            let supports_cooling_profiles = device_props.supports_cooling_profiles.unwrap_or(false);
            let speed_channels = vec!["fan".to_string(), "pump".to_string()];
            for speed_channel in speed_channels {
                channels.insert(
                    speed_channel.to_string(),
                    ChannelInfo {
                        speed_options: Some(SpeedOptions {
                            min_duty: 0,
                            max_duty: 100,
                            profiles_enabled: supports_cooling_profiles,
                            fixed_enabled: true,
                            manual_profiles_enabled: true,
                        }),
                        ..Default::default()
                    },
                );
            }
        }
        if let Some(true) = device_props.supports_lighting {
            let color_channels = vec!["ring".to_string(), "logo".to_string(), "sync".to_string()];
            for channel_name in color_channels {
                let lighting_modes = self.get_color_channel_modes(Some(&channel_name));
                channels.insert(
                    channel_name,
                    ChannelInfo {
                        lighting_modes,
                        ..Default::default()
                    },
                );
            }
            //
        }
        let lighting_speeds = vec![
            "slowest".to_string(),
            "slower".to_string(),
            "normal".to_string(),
            "faster".to_string(),
            "fastest".to_string(),
        ];
        DeviceInfo {
            channels,
            lighting_speeds,
            temp_min: 20,
            temp_max: 60,
            profile_max_length: 9,
            ..Default::default()
        }
    }

    fn get_color_channel_modes(&self, channel_name: Option<&str>) -> Vec<LightingMode> {
        let mut color_modes: Vec<ColorMode> = vec![
            ColorMode::new("off", 0, 0, false, false),
            ColorMode::new("fixed", 1, 1, false, false),
            ColorMode::new("super-fixed", 1, 9, false, false),
            ColorMode::new("fading", 2, 8, true, false),
            ColorMode::new("spectrum-wave", 0, 0, true, true),
            ColorMode::new("breathing", 1, 8, true, false),
            ColorMode::new("super-breathing", 1, 9, true, false),
            ColorMode::new("pulse", 1, 8, true, false),
        ];
        if let Some("ring") = channel_name {
            let mut ring_only_color_modes: Vec<ColorMode> = vec![
                ColorMode::new("marquee-3", 1, 1, true, true),
                ColorMode::new("marquee-4", 1, 1, true, true),
                ColorMode::new("marquee-5", 1, 1, true, true),
                ColorMode::new("marquee-6", 1, 1, true, true),
                ColorMode::new("covering-marquee", 1, 8, true, true),
                ColorMode::new("alternating", 2, 2, true, false),
                ColorMode::new("moving-alternating", 2, 2, true, true),
                ColorMode::new("tai-chi", 2, 2, true, false),
                ColorMode::new("water-cooler", 0, 0, true, false),
                ColorMode::new("loading", 1, 1, true, false),
                ColorMode::new("wings", 1, 1, true, false),
                ColorMode::new("super-wave", 1, 8, true, true),
            ];
            color_modes.append(&mut ring_only_color_modes);
        }
        self.convert_to_channel_lighting_modes(color_modes)
    }
}
